README.md
# Forecasting-Inflation
Forecasting Inflation in unstable environments


Changelog.txt
! For now: keep track of where your data comes from, which packages are needed and on which papers you base your code 
! Options: ### Added, ### Fixed, ### Changed, ### Removed
	- Provide a short description in bullet points

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [1.1.1] - 2023-03-13

### Added

- Get_Data.R to retrieve data
- Based on paper: Forecasting Inflation in a data-rich environment (therefore install github repository)
- Requires packages: fbi, ggplot2, dplyr, tidyverse, TTR, install_github("cykbennie/fbi")
- Inflation plot 

## [1.1.2] - 2023-03-14

### Added

All added R scripts and functions are based on Forecasting Inflation in a data-rich environment
- Models.R:
	- runarx
	- runar
	- runridge
	- runadaelasticnet
	- runrf
- runridge and runadaelasticnet require packages HDeconometrics and glmnet
- runrf require package randomForest

- callModels.R:
	- Calls the models of Models.R
- getRW.R:
	- Gets the forecasts of the random walk model

## [1.1.3] - 2023-03-15

## Added

All added R scripts and functions are based on Forecasting Inflation in a data-rich environment
- Evaluate_forecast.R
	- Needs libraries:
		- tidyverse and Metrics
- Models.R
	- runsvr
	- runsvr requires new package: e1071
	
## [1.1.4] - 2023-03-16	

## Fixed
- Evaluate_forecast.R

## Added
- Models.R
	- Lags of regressors removed from regression matrix
	- Max autoregressive term (p) is set to be 12
		- Everytime the model is estimated (for each t out-of-sample) p is determined	
	- LLF
		- Needs libraries: library(grf)
	- MOTR-BART
		- Needs libraries: install_github("ebprado/MOTR-BART/MOTRbart")
					library(MOTRbart)

## [1.1.5] - 2023-03-17

## Added
- Models.R
	- ic.getShapleyValues
		- Requires: library(kernelshap)
Idea: 
1) get shapley values per time point t: 1 x 10 vector
2) Collect all time points together, Shapley matrix: 277 x 10 matrix
3) Average shapley value: average per column: 1x 10 matrix
4) Plot Shapley value and input x:
	- Store Shapley value per input x : 277 x 1 
	- Store input value : 277 x 1
	- Combine to plot 	

## Changed
- Call_models.R
	- link to Models.R to get shapley values
- Rolling_window.R
	- link to Models.R to get shapley values

getRW.R

#### gets out of sample y and computes random walk forecasts ###
library(roll)
library(dplyr)
load("dataset.Rda")
dates = dataset$date
data = dataset%>%select(-date)%>%as.matrix()
rownames(dataset) = as.character(dates)

nwindows = 277

y = data[,"data.CPIAUCSL"]
y = cbind(y,roll_prod(1+y,3)-1,roll_prod(1+y,6)-1,roll_prod(1+y,12)-1)
yout = tail(y,nwindows)

rw = matrix(NA,nwindows,12)
for(i in 1:12){
  aux=data[(nrow(data)-nwindows-i+1):(nrow(data)-i),"data.CPIAUCSL"]
  rw[,i]=aux;
}

rw3 = tail(embed(y[,2],4)[,4],nwindows)
rw6 = tail(embed(y[,3],7)[,7],nwindows)
rw12 = tail(embed(y[,4],13)[,13],nwindows)
rw = cbind(rw,rw3,rw6,rw12)
colnames(rw) = c(paste("t+",1:12,sep = ""),"acc3","acc6","acc12")

save(yout,file = "yout.rda")
save(rw,file = "rw.rda")

rw_data = data.frame(tail(dates,277), tail(data[,"data.CPIAUCSL"],277), rw[,1])
colnames(rw_data) = c("date", "CPI", "forecast_CPI")
ggplot(forecast_data, aes(x=date)) + 
  geom_line(aes(y = CPI), color = "darkred") + 
  geom_line(aes(y = forecast_CPI), color="steelblue", linetype="twodash")




Models.R
dataprep = function(ind,df,variable, horizon, univar = FALSE)
{
  df=df[ind,]
  y=df[,variable]
  
  if (univar==FALSE){
      x = df
      x = x[, colnames(x) != variable] # Exclude CPI from potential predictors
      max_lag = 1
    } else{
      x = as.matrix(df[,variable])
      x = head(x,-1)
      x = na.omit(x)
      max_lag = 12
  }
  
  # Each row of the resulting matrix consists of x[t], x[t-1],..., x[t-dimension+1]
  # 4 indicates: include the regressors at time t, t-1, t-2 and t-3
  # We only indicate the regressors at time t, but we could extend it here
   X=embed(as.matrix(x), max_lag)
   
   # In case not arp model, we include lag of CPI as regressor
  if (univar == FALSE) {
    y_lag = head(y, -1)
    X = tail(X, (nrow(X) - 1))
    X = cbind(X, y_lag)
  }
  
  Xin=X[-c((nrow(X)-horizon+1):nrow(X)),]
  Xout=X[nrow(X),]
  Xout=t(as.vector(Xout))
  yin=tail(y,nrow(Xin))
  
  return(list(Xin = Xin, Xout = Xout, yin = yin))
  
}





# Benchmark models
# 1: Random walk

# run getRW.R

# 2: ARX(p) model

runarx=function(ind,df,variable,horizon){
  prep_data = dataprep(ind,df,variable,horizon, univar = TRUE)
  ar_terms_in = prep_data$Xin
  yin = prep_data$yin
  ar_terms_out = prep_data$Xout
  
  # First we determine the optimal lag order
  bb=Inf
  best = 1
  for(i in seq(1,ncol(ar_terms_in),1)){
    m=lm(yin~ar_terms_in[,1:i])
    crit=BIC(m)
    if(crit<bb){
      bb=crit
      modelest=m
      best = i
    }
  }
  
  # Second, include regressor terms in regression
  prep_data = dataprep(ind,df,variable,horizon, univar = FALSE)
  x_terms_in = prep_data$Xin
  x_terms_out = prep_data$Xout
  
  ar_n_obs = nrow(ar_terms_in)
  x_n_obs = nrow(x_terms_in)
  dif_obs = x_n_obs - ar_n_obs
  x_terms_in = tail(x_terms_in, x_n_obs - dif_obs)
  
  modelest = lm(yin~ar_terms_in[,1:best]+x_terms_in)
  
  coef=coef(modelest)
  coef[is.na(coef)] = 0
  
  forecast=c(1,ar_terms_out[,1:best], x_terms_out)%*%coef
  
  return(list(forecast=forecast))
}

# Also interesting to compare with "normal" AR(p) model based on BIC:
runar=function(ind,df,variable,horizon){
  prep_data = dataprep(ind,df,variable,horizon, univar = TRUE)
  Xin = prep_data$Xin
  yin = prep_data$yin
  Xout = prep_data$Xout
  
  bb=Inf
  best = 1
  for(i in seq(1,ncol(Xin),1)){
    m=lm(yin~Xin[,1:i])
    crit=BIC(m)
    if(crit<bb){
      bb=crit
      modelest=m
      best = i
    }
  }

  coef=coef(modelest)
  coef[is.na(coef)] = 0
  forecast=c(1,Xout[,1:best])%*%coef
  
  return(list(forecast=forecast))
}

# Machine learning models
# Linear Machine Learning models:

# 1: Ridge regression: alpha = 0 is the ridge penalty
runridge=function(ind,df,variable,horizon, alpha = 0, alpha2 = 0, adaptive = FALSE){
  prep_data = dataprep(ind,df,variable,horizon)
  Xin = prep_data$Xin
  yin = prep_data$yin
  Xout = prep_data$Xout
  
  modelest = ic.glmnet(Xin,yin,alpha = alpha)
  
  # Adaptive ridge (Could be an extension)
  if (adaptive==TRUE){
    cridge = coef(modelest)
    penalty = (abs(cridge[-1])+1/sqrt(length(yin)))^(-1)
    modelest = ic.glmnet(Xin,yin, penalty.factor = penalty, alpha = alpha2)
  }
  
  forecast=predict(modelest,Xout)
  
  if (horizon == 3) {
    shapley_values = ic.getShapleyValues(modelest, Xout, Xin, yin)
  }
  
  ### outputs ###
  coeflvl=coef(modelest)[-1]
  coefpar=coeflvl*apply(Xin,2,sd)
  lambda=modelest$lambda
  outputs=list(coeflvl=coeflvl,coefpar=coefpar,lambda=lambda)
  
  return(list(forecast=forecast, outputs=outputs, shapley_values = shapley_values, Xout = Xout))
}

# 2: Adaptive Elastic Net: alpha = 0.5
runadaelasticnet=function(ind,df,variable,horizon, alpha = 0.5, alpha2 = 0.5, adaptive = TRUE){
  prep_data = dataprep(ind,df,variable,horizon)
  Xin = prep_data$Xin
  yin = prep_data$yin
  Xout = prep_data$Xout
  
  modelest = ic.glmnet(Xin,yin,alpha = alpha)
  
  # Adaptive elastic net
  if (adaptive==TRUE){
    celasticnet = coef(modelest)
    penalty = (abs(celasticnet[-1])+1/sqrt(length(yin)))^(-1)
    modelest = ic.glmnet(Xin,yin, penalty.factor = penalty, alpha = alpha2)
  }
  
  forecast=predict(modelest,Xout)
  if (horizon == 3) {
    shapley_values = ic.getShapleyValues(modelest, Xout, Xin)
  }
  
  ### outputs ###
  coeflvl=coef(modelest)[-1]
  coefpar=coeflvl*apply(Xin,2,sd)
  lambda=modelest$lambda
  outputs=list(coeflvl=coeflvl,coefpar=coefpar,lambda=lambda)
  
  return(list(forecast=forecast, outputs=outputs, shapley_values = shapley_values, Xout = Xout))
}

# Nonlinear ML models
# 1: Random forest
runrf=function(ind,df,variable,horizon){
  prep_data = dataprep(ind,df,variable,horizon)
  Xin = prep_data$Xin
  yin = prep_data$yin
  Xout = prep_data$Xout
  
  modelest=randomForest::randomForest(Xin,yin, importance = TRUE)
  forecast=predict(modelest,Xout)
  
  ## outputs
  importance = randomForest::importance(modelest)
  outputs = list(importance = importance)
  
  return(list(forecast=forecast, outputs = outputs))
}


# 2: SVR
runsvr =function(ind,df,variable,horizon){
  prep_data = dataprep(ind,df,variable,horizon)
  Xin = prep_data$Xin
  yin = prep_data$yin
  Xout = prep_data$Xout
  
  modelest=e1071::svm(Xin,yin, importance = TRUE)
  forecast=predict(modelest,Xout)
  
  return(list(forecast=forecast))
}


# 3: LLF
runllf=function(ind,df,variable,horizon){
  prep_data = dataprep(ind,df,variable,horizon)
  Xin = prep_data$Xin
  yin = prep_data$yin
  Xout = prep_data$Xout
  
  modelest=grf::ll_regression_forest(Xin,yin, num.trees = 10)
  forecast=predict(modelest,Xout)
  
  return(list(forecast=forecast))
}



# 4: MOTR-BART
runmotrbart=function(ind,df,variable,horizon){
  prep_data = dataprep(ind,df,variable,horizon)
  Xin = prep_data$Xin
  yin = prep_data$yin
  Xout = prep_data$Xout
  
  modelest=MOTRbart::motr_bart(Xin,yin, ntrees = 10, nburn = 10, npost = 10)
  forecast = predict_motr_bart(modelest, Xout, 'mean')

  return(list(forecast=forecast))
}

# 5: MRF
runmrf = function(ind,df,variable,horizon){
  prep_data = dataprep(ind,df,variable,horizon)
  Xin = prep_data$Xin
  yin = prep_data$yin
  Xout = prep_data$Xout
  yout = prep_data$yout
  data = cbind(yin, yout, Xin, Xout)
  
  
  return(list(forecast=forecast))
}


accumulate_model = function(forecasts){
  
  acc3 = c(rep(NA,2),sapply(1:(nrow(forecasts)-2), function(x){
    prod(1+diag(forecasts[x:(x+2),1:3]))-1
  })) 
  acc6 = c(rep(NA,5),sapply(1:(nrow(forecasts)-5), function(x){
    prod(1+diag(forecasts[x:(x+5),1:6]))-1
  }))
  acc12 = c(rep(NA,11),sapply(1:(nrow(forecasts)-11), function(x){
    prod(1+diag(forecasts[x:(x+11),1:12]))-1
  }))
  
  forecasts = cbind(forecasts,acc3,acc6,acc12)
  colnames(forecasts) = c(paste("t+",1:12,sep = ""),"acc3","acc6","acc12")
  
  return(forecasts)
  
}

ic.glmnet = function (x, y, crit = c("bic", "aic", "aicc", 
                                     "hqc"), alpha = 1, ...) 
{
  if (is.matrix(x) == FALSE) {
    x = as.matrix(x)
  }
  if (is.vector(y) == FALSE) {
    y = as.vector(y)
  }
  crit = match.arg(crit)
  n = length(y)
  model = glmnet(x = x, y = y, alpha = alpha,...)
  coef = coef(model)
  lambda = model$lambda
  df = model$df
  yhat = cbind(1, x) %*% coef
  residuals = (y - yhat)
  mse = colMeans(residuals^2)
  sse = colSums(residuals^2)
  nvar = df + 1
  bic = n * log(mse) + nvar * log(n)
  aic = n * log(mse) + 2 * nvar
  aicc = aic + (2 * nvar * (nvar + 1))/(n - nvar - 1)
  hqc = n * log(mse) + 2 * nvar * log(log(n))
  sst = (n - 1) * var(y)
  r2 = 1 - (sse/sst)
  adjr2 = (1 - (1 - r2) * (n - 1)/(nrow(x) - nvar - 1))
  crit = switch(crit, bic = bic, aic = aic, aicc = aicc, hqc = hqc)
  selected = best.model = which(crit == min(crit))
  ic = c(bic = bic[selected], aic = aic[selected], aicc = aicc[selected], 
         hqc = hqc[selected])
  result = list(coefficients = coef[, selected], ic = ic, lambda = lambda[selected], 
                nvar = nvar[selected], glmnet = model, residuals = residuals[, 
                                                                             selected], fitted.values = yhat[, selected], ic.range = crit, 
                df = df, call = match.call())
  class(result) = "ic.glmnet"
  return(result)
}

ic.getShapleyValues = function(modelest, Xout, Xin) {
  colnames(Xout) =  c("RPI", "INDPRO", "UNRATE", "HOUST", "DPCERA", "M2SL", "TB3MS", "SP500", "Slope yield", "CPI lag")
  colnames(Xin) = c("RPI", "INDPRO", "UNRATE", "HOUST", "DPCERA", "M2SL", "TB3MS", "SP500", "Slope yield","CPI lag")
  
  system.time(
    shap_lm <- kernelshap(modelest, Xout, Xin)
  )
  shapley_values = shap_lm$S # 1 x 10 vector
  return(shap_lm$S)
}


Get_Data.R
# Get the data from the FredMD macroeconomic database 

# The following packages need to be implemented
# install.packages("stats")
# install.packages("readr")
# install.packages("pracma")

# install.packages("devtools", type = "win.binary")
# library(devtools)
# install_github("cykbennie/fbi")
#install.packages("stargazer")
#library(stargazer)
library(fbi)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(TTR)

# Load the data from the FREDMD database
start_date = "1960-01-01"
data = fredmd("https://files.stlouisfed.org/files/htdocs/fred-md/monthly/current.csv")
data_raw = fredmd("https://files.stlouisfed.org/files/htdocs/fred-md/monthly/current.csv",
                  transform = FALSE)
# Description of all transformations
varlist = fredmd_description
vars = intersect(colnames(data),varlist$fred)

data = data %>% as.tibble()%>%
  select(all_of(c("date",vars)))
varlist = varlist%>%filter(fred%in%vars)
prices_varlist = varlist%>%filter(group=="Prices",tcode==6)

data = data%>% as_tibble()%>%
  select( -all_of(prices_varlist$fred) )

prices = data_raw %>% as_tibble() %>%
  select(all_of(prices_varlist$fred))%>%
  mutate_all(.funs = function(x)100*c(NA,x%>%log()%>%diff()))

data = cbind(data%>%as.data.frame(),prices%>%as.data.frame())

data = data %>%
  filter(date>=start_date)%>%
  select_if(~ !any(is.na(.)))

# CPI : first difference in logarithm. Inflation at month t
CPI = data$CPIAUCSL
date = as.Date(data$date)

# Make the yield data
slope_yield = data_raw$GS10 - data_raw$TB3MS
slope_yield_st = diff(slope_yield)
slope_yield_st = slope_yield_st[12:768]

# Create new dataframe for dates and (transformed) CPI combined
data_CPI = data.frame(
  day = date,
  value = CPI
)

data_CPI_raw = data.frame(
  day = as.Date(data_raw$date),
  value = data_raw$CPIAUCSL
)

# Create time series plot for CPI 
# Time plot for the first differences logarithm: inflation in %
p = ggplot(data_CPI, aes(x=day, y=value)) +
  geom_line() + 
  xlab("") + ylab("Inflation %")

p + scale_x_date(date_breaks = "3 years", date_labels = "%Y")

# Time series plot of CPI before any transformations
# Accumulated inflation therefore an upward trend 
q = ggplot(data_CPI_raw, aes(x=day, y=value)) +
  geom_line() + 
  xlab("") + ylab("Accumulated inflation")

q + scale_x_date(date_breaks = "3 years", date_labels = "%Y")

# Stable and unstable split
# standard deviation out of sample period
sd(CPI[481:757])
# standard deviation for first split
sd(CPI[481:612])
# standard deviation for second split
sd(CPI[613:720])
# standard deviation for third split
sd(CPI[721:757])

save(data,file="data.Rda")

dataset = data.frame(date, data$CPIAUCSL, data$RPI, data$INDPRO, data$UNRATE, data$HOUST,
                data$DPCERA3M086SBEA, data$M2SL, data$TB3MS, data$`S&P 500`, slope_yield_st)
save(dataset, file="dataset.Rda")
#stargazer(dataset[721:757,], type = "latex", title="Descriptive statistics", digits=1, out="table1.txt",  flip=TRUE)

plot(tail(data[,"CPIAUCSL"],180),type = "l")

# Data from https://fredhelp.stlouisfed.org/fred/data/understanding-the-data/recession-bars/
recessions.df = read.table(textConnection(
  "Peak, Trough
1857-06-01, 1858-12-01
1860-10-01, 1861-06-01
1865-04-01, 1867-12-01
1869-06-01, 1870-12-01
1873-10-01, 1879-03-01
1882-03-01, 1885-05-01
1887-03-01, 1888-04-01
1890-07-01, 1891-05-01
1893-01-01, 1894-06-01
1895-12-01, 1897-06-01
1899-06-01, 1900-12-01
1902-09-01, 1904-08-01
1907-05-01, 1908-06-01
1910-01-01, 1912-01-01
1913-01-01, 1914-12-01
1918-08-01, 1919-03-01
1920-01-01, 1921-07-01
1923-05-01, 1924-07-01
1926-10-01, 1927-11-01
1929-08-01, 1933-03-01
1937-05-01, 1938-06-01
1945-02-01, 1945-10-01
1948-11-01, 1949-10-01
1953-07-01, 1954-05-01
1957-08-01, 1958-04-01
1960-04-01, 1961-02-01
1969-12-01, 1970-11-01
1973-11-01, 1975-03-01
1980-01-01, 1980-07-01
1981-07-01, 1982-11-01
1990-07-01, 1991-03-01
2001-03-01, 2001-11-01
2007-12-01, 2009-06-01
2020-02-01, 2020-04-01"), sep=',',
  colClasses=c('Date', 'Date'), header=TRUE)

recessions.trim = subset(recessions.df, Peak >= min(data_CPI$day) )

g = ggplot(data_CPI) + geom_line(aes(x=day, y=value))+ scale_x_date(limits = as.Date(c("1960-01-01","2023-01-01")), date_breaks = "6 years", date_labels = "%Y") + theme_bw()
g = g + geom_rect(data=recessions.trim, aes(xmin=Peak, xmax=Trough, ymin=-Inf, ymax=+Inf), fill='blue', alpha=0.2)


# Out-of-sample start and split in the out-of-sample
h = g + geom_vline(xintercept = as.Date("1999-12-01"), color = "red")+ geom_vline(xintercept = as.Date("2010-12-01"), color = "orange") + geom_vline(xintercept = as.Date("2019-12-01"), color = "orange")+ labs(x = "Date", 
         y = "Inflation %")
ggsave("inflationplot.pdf", plot = h)


Call_models.R

### must add package for specific models ###
#install.packages("devtools")
#install_github("gabrielrvsc/HDeconometrics")
#install.packages("randomForest")
#install.packages("e1071")
#install.packages("grf")
#install_github("ebprado/MOTR-BART/MOTRbart")
#install.packages("kernelshap")
#install.packages("shapviz")
#install_github("philgoucou/macrorf")
library(devtools)
library(HDeconometrics)
library(glmnet)
library(randomForest)
library(e1071)
library(grf)
library(MOTRbart)
library(MacroRF)
library(kernelshap)
library(shapviz)

source('C:/Users/Gebruiker/Documents/Forecasting-Inflation/Analysis/Functions/Rolling_window.R')
source('C:/Users/Gebruiker/Documents/Forecasting-Inflation/Analysis/Functions/Models.R')


#####
## The file with the forecasts will be saved with model_name
model_name = "Elastic net"
## The function called to run models is model_function, which is a function from functions.R
model_function = runadaelasticnet
#####


load("dataset.rda")
dates = dataset$date
data = dataset%>%select(-date)%>%as.matrix()
rownames(data) = as.character(dates)

####### run rolling window ##########
nwindows = 277
model_list = list()
#for(i in 1:12){ 
i=3
  model = rolling_window(model_function,data,nwindows+i-1,i,"data.CPIAUCSL")
  model_list[[i]] = model
  cat(i,"\n")
#}

forecasts = Reduce(cbind,lapply(model_list, function(x)head(x$forecast,nwindows)))

# Get the shapley values of all observations and combine them in a matrix,
# each column represents a regressor
shapley = unlist(lapply(model_list, function(x)x$shapley))
shapley = matrix(shapley, ncol = 10, byrow = TRUE)
avg_shapley = colMeans(shapley)

barplot(height=avg_shapley, names=c("RPI", "INDPRO", "UNRATE", "HOUST", "DPCERA", "M2SL", "TB3MS", "SP500", "Slope yield", "CPI lag"), 
        col="#69b3a2",
        horiz=T
)

Xout = unlist(lapply(model_list, function(x)x$Xout))
Xout = matrix(Xout, ncol = 10, byrow = TRUE)

for (i in 1:ncol(Xout)) {
  plot(Xout[,i], shapley[,i]) # Functional form  
}

# forecasts = accumulate_model(forecasts)

save(forecasts,file = paste("",model_name,".rda",sep = ""))

plot(tail(data[,"data.CPIAUCSL"],277),type = "l")
lines(forecasts[,1],col = 3)

forecast_data = data.frame(tail(dates,277), tail(data[,"data.CPIAUCSL"],277), forecasts[,1])
colnames(forecast_data) = c("date", "CPI", "forecast_CPI")
g = ggplot(forecast_data, aes(x=date)) + 
  geom_line(aes(y = CPI), color = "darkred") + 
  geom_line(aes(y = forecast_CPI), color="steelblue", linetype="twodash")
save(g, file = paste("",model_name,".png",sep = ""))

Rolling_window.R
#fn = function, df = data, nwindow = window size, horizon = forecasting horizon, variable = target variable
rolling_window=function(fn,df,nwindow,horizon,variable,...){
  # nrow(df) = number of observations full sample
  ind=1:nrow(df)
  # window_size = number of observations full sample - the window we want to move over
  window_size=nrow(df)-nwindow
  indmat=matrix(NA,window_size,nwindow)
  indmat[1,]=1:ncol(indmat)
  
  for (i in 2:nrow(indmat)) {
    indmat[i,]=indmat[i-1,]+1
  }
  
  # Apply is a substitute to a loop
  # Margin = 2, manipulation is performed over the columns
  # Fun = fn, the function to apply
  rw = apply(indmat,2,fn,df=df,horizon=horizon,variable=variable,...)
  forecast = unlist(lapply(rw,function(x)x$forecast))
  
  # lapply: performs operations on list objects and returns a list of objects of
  # same length as the original set
  outputs = lapply(rw,function(x)x$outputs)
  
  shapley_values = unlist(lapply(rw, function(x)x$shapley_values))
  Xout = unlist(lapply(rw, function(x)x$Xout))
  return (list(forecast=forecast, outputs=outputs, shapley_values = shapley_values, Xout = Xout))
  
}

Evaluate_forecast.R
#install.packages("Metrics")

library(tidyverse)
library(Metrics)
library(kernelshap)
library(shapviz)

# Extract forecasts from this model:
model_name = "Ridge"
doc_type = ".rda"  
doc_name = paste(model_name,doc_type, sep = "")

# Load the forecasts and actual data
load("dataset.rda")
dates = dataset$date
data = dataset%>%select(-date)%>%as.matrix()
CPI_out = tail(data[, "data.CPIAUCSL"],277)
load(doc_name)
load("yout.rda")

# Columns: t+1,...,t+12, acc 3, acc 6, acc 12
# Time: 277 obs (out-of-sample period)
# For t+12, January 2023 (last data point) should be compared with January 2022
# Compare for period: start sample period: total obs - window until total obs - horizon
nwindows = nrow(forecasts)
rmse_vector = vector( "double" , 15 )
mae_vector = vector( "double" , 15 )
mad_vector = vector( "double" , 15 )


for (i in 1:12) {
  inf_out = tail(CPI_out, nwindows - i)
  forecast_vector = head(forecasts[,i], nwindows - i)
  rmse_vector[i] = rmse(inf_out, forecast_vector)
  mae_vector[i] = mae(inf_out, forecast_vector)
  mad_vector[i] = mad(inf_out, forecast_vector)
}

# Evaluate accumulated forecast
acc3 = yout[,2]
acc3 = tail(acc3, nwindows - 2)
acc6 = yout[,3]
acc6 = tail(acc6, nwindows - 5)
acc12 = yout[,4]
acc12 = tail(acc12, nwindows - 11)

acc3_for = forecasts[,13]
acc3_for = acc3_for[!is.na(acc3_for)]
acc6_for = forecasts[,14]
acc6_for = acc6_for[!is.na(acc6_for)]
acc12_for = forecasts[,15]
acc12_for = acc12_for[!is.na(acc12_for)]


rmse_vector[13] = rmse(acc3, acc3_for)
rmse_vector[14] = rmse(acc6, acc6_for)
rmse_vector[15] = rmse(acc12, acc12_for)

mae_vector[13] = mae(acc3, acc3_for)
mae_vector[14] = mae(acc6, acc6_for)
mae_vector[15] = mae(acc12, acc12_for)

mad_vector[13] = mad(acc3, acc3_for)
mad_vector[14] = mad(acc6, acc6_for)
mad_vector[15] = mad(acc12, acc12_for)

# Store rmse, mae and mad in matrix
metrics = matrix(NA,3,15)
metrics[1,] = rmse_vector
metrics[2,] = mae_vector
metrics[3,] = mad_vector
save(metrics, file = paste("Metrics",model_name,".rda",sep = ""))





